{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Average relative abundances stratified for bile acid in GF and lung transplant/controls\n",
    "\n",
    "Input files: shannon_dv.txt (alpha diversity), otu.txt (OTU table), blastn_alignment.xml (Blastn alignment output), taxonomy.tsv (OTU-to-taxa conversion from Qiime2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "%matplotlib inline\n",
    "import os\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "import matplotlib.patches as mpatches\n",
    "import pickle\n",
    "\n",
    "from matplotlib import transforms, lines, figure\n",
    "import matplotlib.transforms as mtransforms\n",
    "from matplotlib.font_manager import FontProperties\n",
    "from seaborn.utils import remove_na\n",
    "import matplotlib\n",
    "%matplotlib inline\n",
    "matplotlib.rcParams['pdf.fonttype'] = 42\n",
    "matplotlib.rcParams['ps.fonttype'] = 42\n",
    "matplotlib.rcParams['font.family'] = \"sans-serif\"\n",
    "matplotlib.rcParams['font.sans-serif'] = \"Arial\"\n",
    "import warnings\n",
    "warnings.filterwarnings('ignore')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "def setLT_alpha(row):\n",
    "    \"\"\"\n",
    "    Make a column with 'No', 'Yes' which is an snswer to the question 'Did the patient had a lung transplant?'.\n",
    "    Takes row in DF as input and returns value depending on the row value. \n",
    "    \"\"\"\n",
    "    \n",
    "    if row['Lung_Transplant'] == 'yes':\n",
    "            \n",
    "        return 'Lung Transplant'\n",
    "\n",
    "    else:    \n",
    "        return 'Other'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "def setCondition_alpha(row):\n",
    "    \"\"\"\n",
    "    Make a column with 'healthy', 'CF' or 'LT'. Takes row in DF as input and returns value depending on the row \n",
    "    value. \n",
    "    \n",
    "    According to instructions; CF with a LT is classified as healthy.  \n",
    "    \"\"\"\n",
    "    \n",
    "    if row['Lung_Transplant'] == 'yes':\n",
    "        \n",
    "        return 'Lung Transplant'\n",
    "    \n",
    "    elif row['Cystic_fibrosis'] == 'yes':\n",
    "        \n",
    "        return 'Cystic Fibrosis'\n",
    "        \n",
    "    return 'Healthy'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def group_by_heading( some_source ):\n",
    "\n",
    "   buffer= []\n",
    "   for line in some_source:\n",
    "\n",
    "      # Fasta file header, i.e. '>'\n",
    "      if line.startswith( \">\" ):\n",
    "         if buffer: yield buffer\n",
    "         buffer= [ line ]\n",
    "      else:\n",
    "         buffer.append( line )\n",
    "   yield buffer"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def connectFeatureIDtoTaxa(row, taxa_dict, taxa_pick):\n",
    "    \"\"\"\n",
    "    Function which takes a feature ID as input, iterate through a dictionary with feature ID as keys\n",
    "    and taxas as value. \n",
    "    Example on a key/value pair: \n",
    "    '12bf2f325bfc152cf0e6856d8717f82b': \n",
    "    ['k__Bacteria; p__Proteobacteria; c__Alphaproteobacteria; o__Sphingomonadales; f__Sphingomonadaceae', \n",
    "    0.9980158267006052]\n",
    "    \n",
    "    Returns the correspondning taxa.\n",
    "    \"\"\"\n",
    "    for key, value in taxa_dict.items():\n",
    "        \n",
    "        if key == row['#OTU ID']:\n",
    "            \n",
    "            return value[taxa_pick]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "from collections import Counter \n",
    "from itertools import takewhile\n",
    "\n",
    "def get_items_upto_count(dct, n):\n",
    "    \"\"\"\n",
    "    Takes a Counter dict as input and returns the all of the keys with the higest count.\n",
    "    \"\"\"\n",
    "    data = dct.most_common()\n",
    "    val = data[n-1][1] #get the value of n-1th item\n",
    "    #Now collect all items whose value is greater than or equal to `val`.\n",
    "    return list(takewhile(lambda x: x[1] >= val, data))    \n",
    "\n",
    "def most_frequent(List, store_genus): \n",
    "    occurence_count = Counter(List) \n",
    "\n",
    "    ocd = dict(occurence_count)\n",
    "    \n",
    "    # STore all ge+sp which have the highest count\n",
    "    spge_highest_count = get_items_upto_count(occurence_count, 1) # returns [('A', 3), ('C', 3)]\n",
    "    \n",
    "    if len(spge_highest_count) == 1: # One species is the most frequent\n",
    "        return spge_highest_count[0][0]\n",
    "        \n",
    "    else: # All species are equally frequent\n",
    "        \n",
    "        # Multiple species\n",
    "        sp_list = list()\n",
    "        for sp in spge_highest_count:\n",
    "                \n",
    "            # Append with genus\n",
    "            sp_list.append(sp[0].split(' ')[0])\n",
    "            \n",
    "        if len(list(set(sp_list))) == 1:\n",
    "            \n",
    "            genus = list(set(sp_list))[0]\n",
    "            return genus + ' ' + str('spp.')\n",
    "        \n",
    "        else: # Multiple genus\n",
    "            ge_list = list()\n",
    "            for item in spge_highest_count:\n",
    "                ge_list.append(item[0].split(' ')[0]) # Add genuses to list\n",
    "                \n",
    "                if item[0].split(' ')[0] == store_genus:\n",
    "                    return item[0] + ' ' + str('spp.')\n",
    "                    \n",
    "            # No match with Qiime2 genus assignment\n",
    "                    \n",
    "            # Find most abundant genus in ge_list\n",
    "            ge_count = Counter(ge_list)\n",
    "                    \n",
    "            # Store genus with the highest count\n",
    "            ge_highest_count = get_items_upto_count(ge_count, 1) # returns [('A', 3), ('C', 3)]\n",
    "        \n",
    "            return ge_highest_count[0][0] + ' ' + str('spp.')\n",
    "                     \n",
    "\n",
    "def mostIdenticalGenus(row, dict_input):\n",
    "    \"\"\"\n",
    "    Row and OTUid-BLASTalignments dictionary as input. Depending on what Qiime2 assigned as taxa, I return the \n",
    "    most matching taxa from the BLAST aligments. If nothing (genus and/or specie) is identical in BLAST and Qiime2\n",
    "    I return the BLAST assignment which got the most occuring assignments. \n",
    "    \n",
    "    \"\"\"\n",
    "\n",
    "    # THere are some alignments with identical filter values but where one is associated with human, CF \n",
    "    # and/or lungs (manually curated) and the other with something else (plants for example)\n",
    "    # Key is BLAST alignment name, value is manually curated. \n",
    "    manually_curated_taxa_dict = {'Stenotrophomonas pavanii':'Stenotrophomonas maltophilia', \n",
    "                  'Ralstonia solanacearum':'Ralstonia mannitolilytica'}\n",
    "    \n",
    "    # If OTU in dict keys\n",
    "    if row['#OTU ID'] in dict_input:\n",
    "        \n",
    "        store_genus_list = list()\n",
    "        \n",
    "        # Go through the value lists for that key\n",
    "        for val in dict_input[row['#OTU ID']]:\n",
    "            \n",
    "            # If the assigned BLAST genus+specie match the Qiime2 Genus+Specie, pick that BLAST assignment\n",
    "            if val[1].split(' ')[0] == row['Genus'] and val[1].split(' ')[1] == row['Species']:\n",
    "                \n",
    "                # Check against the manually curated taxa\n",
    "                if val[1] in manually_curated_taxa_dict:\n",
    "                    \n",
    "                    return manually_curated_taxa_dict[val[1]]\n",
    "                \n",
    "                else:\n",
    "                    return val[1]\n",
    "                \n",
    "            # If the assigned BLAST genus match the Qiime2 Genus, store that BLAST assignment\n",
    "            elif val[1].split(' ')[0] == row['Genus']:\n",
    "                store_genus_list.append('Genus')\n",
    "            \n",
    "        if list(set(store_genus_list)) != 0:\n",
    "            \n",
    "            store_genus = str(list(set(store_genus_list)))\n",
    "            \n",
    "        else:\n",
    "        \n",
    "            store_genus = 'No Genus'\n",
    "                \n",
    "        # Create DF from the key values, example value in dict looks like this:\n",
    "        #[[' Moraxella nonliquefaciens strain 4663/62 16S ribosomal RNA, partial sequence','Moraxella nonliquefaciens', 5.05092e-87, 99.429], \n",
    "        #[' Moraxella nonliquefaciens strain CCUG 348 16S ribosomal RNA, partial sequence', 'Moraxella nonliquefaciens', 5.05092e-87, 99.429], \n",
    "        #[' Moraxella catarrhalis strain Ne 11 16S ribosomal RNA, partial sequence', 'Moraxella catarrhalis', 5.05092e-87, 99.429]]\n",
    "        value_df = pd.DataFrame(dict_input[row['#OTU ID']], columns = ['Full name', 'Short name', 'e-value', 'ident.%'])\n",
    "        \n",
    "        kk = value_df.groupby('e-value')\n",
    "        \n",
    "        # Select the top group (with lowest e-value) and make 'short list' column to list\n",
    "        shoart_name_list = kk.get_group((list(kk.groups)[0]))['Short name'].tolist()\n",
    "        \n",
    "        # Check for the most abundant short name in list   \n",
    "        most_freq_short_name = most_frequent(shoart_name_list, store_genus)\n",
    "        \n",
    "        # Return the most abundant BLAST genus+specie\n",
    "        # Check against manually curated dict\n",
    "        if most_freq_short_name in manually_curated_taxa_dict:\n",
    "            \n",
    "            return manually_curated_taxa_dict[most_freq_short_name]\n",
    "        \n",
    "        else:\n",
    "            return most_freq_short_name\n",
    "                \n",
    "                \n",
    "    # If OTU ID can't be found in BLAST dict, take what has been assigned in Qiime2\n",
    "    elif (row['Genus'] != 'Undetermined') & (row['Species'] != 'Undetermined'): # Genus + species are available\n",
    "        \n",
    "        return str(row['Genus']) + ' ' + str(row['Species'])\n",
    "    \n",
    "    elif (row['Genus'] != 'Undetermined') & (row['Species'] == 'Undetermined'): # Genus is available\n",
    "        \n",
    "        return str(row['Genus']) + ' ' + str('spp.')\n",
    "\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import df with Shannon diveristy\n",
    "\n",
    "shannon_meta = pd.read_csv('shannon_dv.txt', \n",
    "                     sep = '\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(827, 66)\n",
      "(825, 66)\n"
     ]
    }
   ],
   "source": [
    "# Remove samID 'dup05-013-1B' bc it is a duplicated samID\n",
    "print(shannon_meta.shape)\n",
    "\n",
    "shannon_meta = shannon_meta[shannon_meta.PatientID != 'dup05-013-1']\n",
    "\n",
    "print(shannon_meta.shape)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Make a string out of the nans\n",
    "shannon_meta['Lung_Transplant'].fillna('NA', inplace = True)\n",
    "shannon_meta['Cystic_fibrosis'].fillna('NA', inplace = True)\n",
    "shannon_meta['Swab_site'].fillna('NA', inplace = True)\n",
    "shannon_meta['Swab_occasion'].fillna('NA', inplace = True)\n",
    "shannon_meta['Total Bile Acid BAL'].fillna('NA', inplace = True)\n",
    "shannon_meta['Total Bile Acid GF'].fillna('NA', inplace = True)\n",
    "shannon_meta[['GES_combo']] = shannon_meta[['GES_combo']].astype(str)\n",
    "\n",
    "shannon_meta['Condition'] = shannon_meta.apply (lambda row: setCondition_alpha(row), axis=1)\n",
    "shannon_meta['LT_plot'] = shannon_meta.apply (lambda row: setLT_alpha(row), axis = 1)\n",
    "\n",
    "# Remove the CF patients, but keep all PPI\n",
    "shannon_meta_initial = shannon_meta[((shannon_meta['Swab_site'] == 'lung') |\n",
    "                                (shannon_meta['Swab_site'] == 'stomach') |\n",
    "                                (shannon_meta['Swab_site'] == 'throat')) &\n",
    "                             ((shannon_meta['Swab_occasion'] == 'initial') | (shannon_meta['Swab_occasion'] == 'NA')| (shannon_meta['Swab_occasion'] == 'PPI_final')) &\n",
    "                               ((shannon_meta['Condition'] != 'Cystic fibrosis'))]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "lung\n",
      "stomach\n",
      "throat\n",
      "3\n"
     ]
    }
   ],
   "source": [
    "# Find the samID which have high/low bile acid in GF\n",
    "swab_sit = ['lung', 'stomach', 'throat']\n",
    "\n",
    "LT_random_groups = list()\n",
    "\n",
    "for site in swab_sit:\n",
    "    print(site)\n",
    "    smh = shannon_meta_initial[(shannon_meta_initial['Swab_site'] == site) &\n",
    "                                (shannon_meta_initial['Total Bile Acid GF'] != 'NA')]\n",
    "\n",
    "    LT_random_groups.append(smh)\n",
    "            \n",
    "print(len(LT_random_groups))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [],
   "source": [
    "# What are the samIDs in the different groups?\n",
    "LT_random_groups_samID = []\n",
    "\n",
    "for groups in LT_random_groups:\n",
    "    \n",
    "    samID = groups['SamID'].tolist()\n",
    "    \n",
    "    LT_random_groups_samID.append(samID)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the normalized OTU table\n",
    "\n",
    "OTUtable = pd.read_csv('otu.txt', \n",
    "                      sep = '\\t', index_col = 0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Normalization within each sample\n",
    "#-------------------------------------\n",
    "otu_table_norm = OTUtable.copy()\n",
    "\n",
    "\n",
    "# Calculate sum of each column\n",
    "otu_col_sum = list()\n",
    "\n",
    "for item in OTUtable.columns:\n",
    "    total = OTUtable[item].sum()\n",
    "    \n",
    "    otu_col_sum.append(total)\n",
    "\n",
    "# Divide each row in otu_table with the sum of each column to get the percentage\n",
    "otu_table_norm = otu_table_norm.div(otu_col_sum, axis=1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Get the columns from OTUtable within each group \n",
    "\n",
    "LT_random_groups_OTUs = []\n",
    "\n",
    "for group_list in LT_random_groups_samID:\n",
    "    \n",
    "    OTUdf = otu_table_norm[group_list]\n",
    "    \n",
    "    LT_random_groups_OTUs.append(OTUdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                         Feature ID  \\\n",
      "0  0001d123420b59585627edf5a1292ae8   \n",
      "1  0004ec6a75f43f63758f414c90dabd77   \n",
      "2  000c3b583371be35cc4187f81dfbb602   \n",
      "3  000db46fcff4701589d112bc2a6044de   \n",
      "4  00186d4a03643fc5458d8a2be2430598   \n",
      "\n",
      "                                               Taxon  Confidence  \n",
      "0  k__Bacteria; p__Proteobacteria; c__Alphaproteo...    0.998325  \n",
      "1  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...    0.936442  \n",
      "2  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...    0.995499  \n",
      "3  k__Bacteria; p__Firmicutes; c__Clostridia; o__...    0.797533  \n",
      "4  k__Bacteria; p__Firmicutes; c__Clostridia; o__...    0.921369  \n"
     ]
    }
   ],
   "source": [
    "# Load the OTU ID-to-taxa conversion table\n",
    "\n",
    "taxa_import = pd.read_csv('taxonomy.tsv',\n",
    "                         sep = '\\t')\n",
    "print(taxa_import.head())\n",
    "\n",
    "taxa_dict = taxa_import.set_index('Feature ID').T.to_dict('list')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "# I want to replace the feature ID with the taxa found in taxa_table\n",
    "\n",
    "taxa_pick = 0\n",
    "\n",
    "taxas = ['Kingdom','Phylum','Class','Order','Family','Genus','Species']\n",
    "\n",
    "# Reset index to get #OTU ID as a column\n",
    "for group in LT_random_groups_OTUs:\n",
    "    \n",
    "    group.reset_index(inplace = True)\n",
    "\n",
    "    # Add taxa to melted index/OTU ID table\n",
    "\n",
    "    group['#Taxa'] = group.apply(connectFeatureIDtoTaxa, args=(taxa_dict, taxa_pick), axis = 1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "for group in LT_random_groups_OTUs:\n",
    "    # To remove the '[' and ']' some taxa names have in them\n",
    "    group['#Taxa'].replace('\\[','', regex=True, inplace=True)\n",
    "    group['#Taxa'].replace('\\]','', regex=True, inplace=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "LT_random_groups_OTUs_expanded = []\n",
    "\n",
    "for group in LT_random_groups_OTUs:\n",
    "    \n",
    "    # Split column '#Taxa' into separate taxas\n",
    "    OTUmetamerge_expanded = group.join(group['#Taxa'].str.split(';', expand=True).add_prefix('Taxon').fillna(np.nan))\n",
    "\n",
    "    # Set column names\n",
    "    OTUmetamerge_expanded = OTUmetamerge_expanded.rename(columns = {'Taxon0':'Kingdom', \n",
    "                                                               'Taxon1':'Phylum',\n",
    "                                                               'Taxon2':'Class',\n",
    "                                                               'Taxon3':'Order',\n",
    "                                                               'Taxon4':'Family',\n",
    "                                                               'Taxon5':'Genus',\n",
    "                                                               'Taxon6':'Species'})\n",
    "\n",
    "    # Can I remove the 'D_x__' pattern from the taxa columns?\n",
    "    OTUmetamerge_expanded['Kingdom'] = OTUmetamerge_expanded['Kingdom'].str[3:]\n",
    "    OTUmetamerge_expanded['Phylum'] = OTUmetamerge_expanded['Phylum'].str[4:]\n",
    "    OTUmetamerge_expanded['Class'] = OTUmetamerge_expanded['Class'].str[4:]\n",
    "    OTUmetamerge_expanded['Order'] = OTUmetamerge_expanded['Order'].str[4:]\n",
    "    OTUmetamerge_expanded['Family'] = OTUmetamerge_expanded['Family'].str[4:]\n",
    "    OTUmetamerge_expanded['Genus'] = OTUmetamerge_expanded['Genus'].str[4:]\n",
    "    OTUmetamerge_expanded['Species'] = OTUmetamerge_expanded['Species'].str[4:]\n",
    "    \n",
    "    # Fill blanks with NaN\n",
    "    OTUmetamerge_expanded.replace(r'^\\s*$', np.nan, regex=True, inplace = True)\n",
    "    \n",
    "    # Replace NaN to 'Undetermined'\n",
    "    OTUmetamerge_expanded.fillna('Undetermined', inplace = True)\n",
    "    \n",
    "    # Add df in list with the other groups\n",
    "    LT_random_groups_OTUs_expanded.append(OTUmetamerge_expanded)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Fastq files run in BLASTn 16S database and collected the top 10 query matches in a XML single file.\n",
    "\n",
    "Output is a XML file with all the alignments."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Blast import NCBIXML\n",
    "\n",
    "# Used the 16S database in NCBI to do this search\n",
    "\n",
    "xml_path = 'blastn_alignment.xml'\n",
    "\n",
    "E_VALUE_THRESH = 0.01\n",
    " \n",
    "# To store results per query in dict with OTU ID as key and alignment info as values in lists\n",
    "store_dicttBIG = {}\n",
    "\n",
    "# Open and parse XML file\n",
    "for blast_record in NCBIXML.parse(open(xml_path)):\n",
    "    \n",
    "    for alignment in blast_record.alignments:\n",
    "        for hsp in alignment.hsps:\n",
    "            if hsp.expect < E_VALUE_THRESH:\n",
    "                \n",
    "                # Get the % identity\n",
    "                pindent = float(\"%0.3f\" % (100 * float(hsp.identities) / float(hsp.align_length)))\n",
    "                \n",
    "                # Genus + species name\n",
    "                ge_sp_name = ' '.join([alignment.title.split('|')[4].split(' ')[1], alignment.title.split('|')[4].split(' ')[2]])\n",
    "            \n",
    "                # Sequence title\n",
    "                seq_title = alignment.title.split('|')[4]\n",
    "            \n",
    "                # e-value\n",
    "                e_val = hsp.expect\n",
    "                \n",
    "                # Do some filtering on the % identity and e value\n",
    "                ident_threshold = float(99)\n",
    "                e_val_threshold = float(1e-50)\n",
    "                \n",
    "                if pindent > ident_threshold and e_val < e_val_threshold:\n",
    "                    \n",
    "                    align_info = [seq_title, ge_sp_name, e_val, pindent]\n",
    "            \n",
    "                    # Check if key exists in dict and append new value if it exits\n",
    "                    if blast_record.query in store_dicttBIG:\n",
    "                        store_dicttBIG[blast_record.query].append(align_info)\n",
    "                \n",
    "                    # Create new key in dict if not already there\n",
    "                    else:\n",
    "                        store_dicttBIG[blast_record.query] = [align_info]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                            #OTU ID  04-050-1B  04-064-5B  04-066-0B  \\\n",
      "0  0001d123420b59585627edf5a1292ae8        0.0        0.0        0.0   \n",
      "1  0004ec6a75f43f63758f414c90dabd77        0.0        0.0        0.0   \n",
      "2  000c3b583371be35cc4187f81dfbb602        0.0        0.0        0.0   \n",
      "3  000db46fcff4701589d112bc2a6044de        0.0        0.0        0.0   \n",
      "4  00186d4a03643fc5458d8a2be2430598        0.0        0.0        0.0   \n",
      "\n",
      "   04-067-2B  04-076-5B  04-077-1B  04-079-6B  04-081-3B  04-084-1B  ...  \\\n",
      "0        0.0        0.0        0.0        0.0        0.0        0.0  ...   \n",
      "1        0.0        0.0        0.0        0.0        0.0        0.0  ...   \n",
      "2        0.0        0.0        0.0        0.0        0.0        0.0  ...   \n",
      "3        0.0        0.0        0.0        0.0        0.0        0.0  ...   \n",
      "4        0.0        0.0        0.0        0.0        0.0        0.0  ...   \n",
      "\n",
      "   K05-387-1B                                              #Taxa   Kingdom  \\\n",
      "0         0.0  k__Bacteria; p__Proteobacteria; c__Alphaproteo...  Bacteria   \n",
      "1         0.0  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...  Bacteria   \n",
      "2         0.0  k__Bacteria; p__Bacteroidetes; c__Bacteroidia;...  Bacteria   \n",
      "3         0.0  k__Bacteria; p__Firmicutes; c__Clostridia; o__...  Bacteria   \n",
      "4         0.0  k__Bacteria; p__Firmicutes; c__Clostridia; o__...  Bacteria   \n",
      "\n",
      "           Phylum                Class          Order             Family  \\\n",
      "0  Proteobacteria  Alphaproteobacteria    Rhizobiales  Bradyrhizobiaceae   \n",
      "1   Bacteroidetes          Bacteroidia  Bacteroidales     Prevotellaceae   \n",
      "2   Bacteroidetes          Bacteroidia  Bacteroidales     Prevotellaceae   \n",
      "3      Firmicutes           Clostridia  Clostridiales       Undetermined   \n",
      "4      Firmicutes           Clostridia  Clostridiales    Lachnospiraceae   \n",
      "\n",
      "          Genus       Species    BLAST Genus.Specie  \n",
      "0   Balneimonas  Undetermined  Microvirga aerophila  \n",
      "1    Prevotella  Undetermined   Prevotella buccalis  \n",
      "2    Prevotella  Undetermined       Prevotella spp.  \n",
      "3  Undetermined  Undetermined                  None  \n",
      "4  Undetermined  Undetermined                  None  \n",
      "\n",
      "[5 rows x 33 columns]\n"
     ]
    }
   ],
   "source": [
    "# Return the BLAST genus + specie values to the mono_mulit_groups_expanded's items\n",
    "# Only for the rows where taxa wasn't assigned by Qiime2 = these are the OTU keys stored in OTU_BLAST_dict\n",
    "\n",
    "for groups in LT_random_groups_OTUs_expanded:\n",
    "    \n",
    "    groups['BLAST Genus.Specie'] = groups.apply(lambda row: mostIdenticalGenus(row, store_dicttBIG), axis = 1)\n",
    "    \n",
    "print(LT_random_groups_OTUs_expanded[0].head())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#OTU ID</th>\n",
       "      <th>04-064-5T</th>\n",
       "      <th>04-066-0T</th>\n",
       "      <th>04-067-2T</th>\n",
       "      <th>04-076-5T</th>\n",
       "      <th>04-077-1T</th>\n",
       "      <th>04-079-6T</th>\n",
       "      <th>04-081-3T</th>\n",
       "      <th>04-084-1T</th>\n",
       "      <th>04-115-7T</th>\n",
       "      <th>...</th>\n",
       "      <th>Phylum</th>\n",
       "      <th>Class</th>\n",
       "      <th>Order</th>\n",
       "      <th>Family</th>\n",
       "      <th>Genus</th>\n",
       "      <th>Species</th>\n",
       "      <th>BLAST Genus.Specie</th>\n",
       "      <th>BLAST Genus</th>\n",
       "      <th>BLAST Specie</th>\n",
       "      <th>Full Taxa</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0001d123420b59585627edf5a1292ae8</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Proteobacteria</td>\n",
       "      <td>Alphaproteobacteria</td>\n",
       "      <td>Rhizobiales</td>\n",
       "      <td>Bradyrhizobiaceae</td>\n",
       "      <td>Balneimonas</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Microvirga aerophila</td>\n",
       "      <td>Microvirga</td>\n",
       "      <td>aerophila</td>\n",
       "      <td>Proteobacteria.Alphaproteobacteria.Rhizobiales...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>0004ec6a75f43f63758f414c90dabd77</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Prevotellaceae</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Prevotella buccalis</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>buccalis</td>\n",
       "      <td>Bacteroidetes.Bacteroidia.Bacteroidales.Prevot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>000c3b583371be35cc4187f81dfbb602</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Bacteroidetes</td>\n",
       "      <td>Bacteroidia</td>\n",
       "      <td>Bacteroidales</td>\n",
       "      <td>Prevotellaceae</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Prevotella spp.</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>spp.</td>\n",
       "      <td>Bacteroidetes.Bacteroidia.Bacteroidales.Prevot...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>000db46fcff4701589d112bc2a6044de</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Clostridiales</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>None</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Firmicutes.Clostridia.Clostridiales.Undetermin...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>00186d4a03643fc5458d8a2be2430598</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>Firmicutes</td>\n",
       "      <td>Clostridia</td>\n",
       "      <td>Clostridiales</td>\n",
       "      <td>Lachnospiraceae</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>None</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Undetermined</td>\n",
       "      <td>Firmicutes.Clostridia.Clostridiales.Lachnospir...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 41 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                            #OTU ID  04-064-5T  04-066-0T  04-067-2T  \\\n",
       "0  0001d123420b59585627edf5a1292ae8        0.0        0.0        0.0   \n",
       "1  0004ec6a75f43f63758f414c90dabd77        0.0        0.0        0.0   \n",
       "2  000c3b583371be35cc4187f81dfbb602        0.0        0.0        0.0   \n",
       "3  000db46fcff4701589d112bc2a6044de        0.0        0.0        0.0   \n",
       "4  00186d4a03643fc5458d8a2be2430598        0.0        0.0        0.0   \n",
       "\n",
       "   04-076-5T  04-077-1T  04-079-6T  04-081-3T  04-084-1T  04-115-7T  ...  \\\n",
       "0        0.0        0.0        0.0        0.0        0.0        0.0  ...   \n",
       "1        0.0        0.0        0.0        0.0        0.0        0.0  ...   \n",
       "2        0.0        0.0        0.0        0.0        0.0        0.0  ...   \n",
       "3        0.0        0.0        0.0        0.0        0.0        0.0  ...   \n",
       "4        0.0        0.0        0.0        0.0        0.0        0.0  ...   \n",
       "\n",
       "           Phylum                Class          Order             Family  \\\n",
       "0  Proteobacteria  Alphaproteobacteria    Rhizobiales  Bradyrhizobiaceae   \n",
       "1   Bacteroidetes          Bacteroidia  Bacteroidales     Prevotellaceae   \n",
       "2   Bacteroidetes          Bacteroidia  Bacteroidales     Prevotellaceae   \n",
       "3      Firmicutes           Clostridia  Clostridiales       Undetermined   \n",
       "4      Firmicutes           Clostridia  Clostridiales    Lachnospiraceae   \n",
       "\n",
       "          Genus       Species    BLAST Genus.Specie   BLAST Genus  \\\n",
       "0   Balneimonas  Undetermined  Microvirga aerophila    Microvirga   \n",
       "1    Prevotella  Undetermined   Prevotella buccalis    Prevotella   \n",
       "2    Prevotella  Undetermined       Prevotella spp.    Prevotella   \n",
       "3  Undetermined  Undetermined                  None  Undetermined   \n",
       "4  Undetermined  Undetermined                  None  Undetermined   \n",
       "\n",
       "   BLAST Specie                                          Full Taxa  \n",
       "0     aerophila  Proteobacteria.Alphaproteobacteria.Rhizobiales...  \n",
       "1      buccalis  Bacteroidetes.Bacteroidia.Bacteroidales.Prevot...  \n",
       "2          spp.  Bacteroidetes.Bacteroidia.Bacteroidales.Prevot...  \n",
       "3  Undetermined  Firmicutes.Clostridia.Clostridiales.Undetermin...  \n",
       "4  Undetermined  Firmicutes.Clostridia.Clostridiales.Lachnospir...  \n",
       "\n",
       "[5 rows x 41 columns]"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "for groups in LT_random_groups_OTUs_expanded:\n",
    "    \n",
    "    # Split BLAST Genus.Specie column into two, one with Genus name and one with specie name\n",
    "    groups[['BLAST Genus','BLAST Specie']] = groups['BLAST Genus.Specie'].str.split(\" \", n= 1, expand=True)\n",
    "    \n",
    "    # Fill None cells with 'Undetermined'\n",
    "    groups[['BLAST Genus','BLAST Specie']] = groups[['BLAST Genus','BLAST Specie']].fillna('Undetermined')\n",
    "\n",
    "    # Join all Taxa Level Names with dot\n",
    "    groups['Full Taxa'] = groups['Phylum'] + '.' + groups['Class'] + '.' + groups['Order'] + '.' + groups['Family'] + '.' + groups['BLAST Genus'] #+ '.' + groups['BLAST Specie']\n",
    "    \n",
    "    # If a 'Full Taxa' == Undetermined Undetermined, I wanna change that to just one 'Undetermined'\n",
    "    groups.loc[groups['Full Taxa'] == 'Undetermined Undetermined', 'Full Taxa'] = 'Undetermined'\n",
    "    \n",
    "LT_random_groups_OTUs_expanded[2].head()\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Plotting starts..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "scrolled": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "23\n",
      "32\n",
      "28\n",
      "       index                                          Full Taxa  \\\n",
      "0  04-064-5T  Actinobacteria.Actinobacteria.Actinomycetales....   \n",
      "1  04-064-5T  Bacteroidetes.Bacteroidia.Bacteroidales.Parapr...   \n",
      "2  04-064-5T  Bacteroidetes.Bacteroidia.Bacteroidales.Porphy...   \n",
      "3  04-064-5T  Bacteroidetes.Bacteroidia.Bacteroidales.Prevot...   \n",
      "4  04-064-5T  Firmicutes.Bacilli.Gemellales.Gemellaceae.Gemella   \n",
      "\n",
      "   value percentage  \n",
      "0          0.020396  \n",
      "1          0.074695  \n",
      "2          0.050284  \n",
      "3          0.076634  \n",
      "4          0.014272  \n"
     ]
    }
   ],
   "source": [
    "samID_list_str = ['lung', \n",
    "                  'stomach', \n",
    "                  'throat']\n",
    "\n",
    "taxa_BLAST_bar_plotting_ALL = []\n",
    "\n",
    "percentage = 0.01\n",
    "\n",
    "for i in range(len(LT_random_groups_samID)):\n",
    "    \n",
    "    # To remove the '[' and ']' some taxa names have in them\n",
    "    LT_random_groups_OTUs_expanded[i]['Full Taxa'].replace('\\[','', regex=True, inplace=True)\n",
    "    LT_random_groups_OTUs_expanded[i]['Full Taxa'].replace('\\]','', regex=True, inplace=True)\n",
    "    \n",
    "    \n",
    "    # Remove the coloumns I do not want when doing bar plots\n",
    "    tmp = LT_random_groups_OTUs_expanded[i].drop(['#OTU ID', \n",
    "                       '#Taxa', \n",
    "                       'Kingdom', \n",
    "                       'Phylum', \n",
    "                       'Class', \n",
    "                       'Order', \n",
    "                       'Family', \n",
    "                       'Genus', \n",
    "                       'Species', \n",
    "                       'BLAST Genus.Specie', \n",
    "                       'BLAST Genus', \n",
    "                       'BLAST Specie'], axis = 1)\n",
    "    \n",
    "    tmp = tmp.set_index('Full Taxa')\n",
    "    \n",
    "    # Transpose dataframe (sampleID as rows and TAXAs as columns)\n",
    "    tmp_transpose = tmp.T\n",
    "    tmp_transpose.reset_index(inplace=True)\n",
    "    \n",
    "    # Make OTU transposed table tidy! :)\n",
    "    tmp_table_melt = pd.melt(tmp_transpose, id_vars=['index'])\n",
    "\n",
    "    # Remove rows with value > 0\n",
    "    tmp_table_melt = tmp_table_melt[tmp_table_melt.value != 0]\n",
    "    \n",
    "    # Groupby index and Taxa and sum the percentages of each group\n",
    "    taxa_val_per = tmp_table_melt.groupby(['index', 'Full Taxa'])['value'].agg('sum').reset_index()\n",
    "    \n",
    "    # change name of the 'value' column to know that it is percentage now\n",
    "    taxa_val_per.columns = ['index', 'Full Taxa', 'value percentage'] \n",
    "    \n",
    "    # Only keep values over a certain value percentage\n",
    "    taxa_val_per2 = taxa_val_per[taxa_val_per['value percentage'] > percentage]\n",
    "    \n",
    "    taxa_val_per2.reset_index(inplace=True, drop=True)\n",
    "    \n",
    "    ############# This is to get every taxon < percentage specified above to be grouped together and presented as\n",
    "    ############# an 'Other' taxon. To get 100% for all samples. \n",
    "    # Get the index in a list\n",
    "    taxa_val_per2_index_list = list(set(taxa_val_per2['index'].tolist()))\n",
    "    taxa_val_per2_index_list.sort()\n",
    "    \n",
    "    # Sum the value percentage for each index and put in list\n",
    "    sumed_val_percent_list = taxa_val_per2.groupby('index').sum()['value percentage'].tolist()\n",
    "    \n",
    "    # Go through the list to add a row with 'other' (everything which has pecentage < percentage (see above) is summarized as 'other')\n",
    "    for subID in range(len(taxa_val_per2_index_list)):\n",
    "        taxa_val_per2.loc[taxa_val_per2.shape[0] + subID] = [taxa_val_per2_index_list[subID]] + ['Other'] + [1-sumed_val_percent_list[subID]]  \n",
    "        \n",
    "    # Store dfs in list\n",
    "    print(len(list(set(taxa_val_per2['index'].tolist()))))\n",
    "    taxa_BLAST_bar_plotting_ALL.append(taxa_val_per2)\n",
    "\n",
    "print(taxa_BLAST_bar_plotting_ALL[2].head())\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Use the same color dict as was created in \"2021-05-03.Average_relAbund_GES_GITHUB\"\n",
    "\n",
    "color_dict_file = 'color_dict.pkl'\n",
    "\n",
    "with open(color_dict_file, 'rb') as f:\n",
    "    test = pickle.load(f)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# To get a concensus for the group with high/low bile acid, LT vs non-LT"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                    value percentage  \\\n",
      "Full Taxa                                                              \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...          0.011062   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Bactero...          0.014901   \n",
      "Firmicutes.Bacilli.Bacillales.Bacillaceae.Bacillus          0.275221   \n",
      "Firmicutes.Bacilli.Bacillales.Staphylococcaceae...          0.967110   \n",
      "Firmicutes.Bacilli.Lactobacillales.Leuconostoca...          0.083186   \n",
      "\n",
      "                                                    mean rel abund  index  \n",
      "Full Taxa                                                                  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...        0.005577      1  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Bactero...        0.007513      1  \n",
      "Firmicutes.Bacilli.Bacillales.Bacillaceae.Bacillus        0.138761      1  \n",
      "Firmicutes.Bacilli.Bacillales.Staphylococcaceae...        0.487598      1  \n",
      "Firmicutes.Bacilli.Lactobacillales.Leuconostoca...        0.041941      1  \n",
      "                                                    value percentage  \\\n",
      "Full Taxa                                                              \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...          0.011430   \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...          0.013919   \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...          0.022811   \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...          0.013088   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...          0.013198   \n",
      "\n",
      "                                                    mean rel abund  index  \n",
      "Full Taxa                                                                  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...        0.005919      1  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...        0.007207      2  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...        0.011811      2  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...        0.006777      1  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...        0.006834      1  \n",
      "                                                    value percentage  \\\n",
      "Full Taxa                                                              \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.C...          0.014871   \n",
      "Firmicutes.Bacilli.Bacillales.Staphylococcaceae...          0.019411   \n",
      "Firmicutes.Bacilli.Gemellales.Gemellaceae.Gemella           0.028493   \n",
      "Firmicutes.Bacilli.Lactobacillales.Streptococca...          0.934955   \n",
      "Other                                                       0.002270   \n",
      "\n",
      "                                                    mean rel abund  index  \n",
      "Full Taxa                                                                  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.C...        0.014871      1  \n",
      "Firmicutes.Bacilli.Bacillales.Staphylococcaceae...        0.019411      1  \n",
      "Firmicutes.Bacilli.Gemellales.Gemellaceae.Gemella         0.028493      1  \n",
      "Firmicutes.Bacilli.Lactobacillales.Streptococca...        0.934955      1  \n",
      "Other                                                     0.002270      1  \n",
      "                                                    value percentage  \\\n",
      "Full Taxa                                                              \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...          0.019194   \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...          0.102510   \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.U...          0.023331   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Bactero...          0.233981   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...          0.044455   \n",
      "\n",
      "                                                    mean rel abund  index  \n",
      "Full Taxa                                                                  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...        0.006373      3  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...        0.034035      1  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.U...        0.007746      1  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Bactero...        0.077686      5  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...        0.014760      8  \n",
      "                                                    value percentage  \\\n",
      "Full Taxa                                                              \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...          0.012781   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Bactero...          0.103448   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...          0.089579   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Prevote...          0.254494   \n",
      "Cyanobacteria.Chloroplast.Streptophyta.Undeterm...          0.040983   \n",
      "\n",
      "                                                    mean rel abund  index  \n",
      "Full Taxa                                                                  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...        0.007568      2  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Bactero...        0.061253      1  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...        0.053040      1  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Prevote...        0.150689      4  \n",
      "Cyanobacteria.Chloroplast.Streptophyta.Undeterm...        0.024267      1  \n",
      "                                                    value percentage  \\\n",
      "Full Taxa                                                              \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...          0.038057   \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...          0.018770   \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...          0.094879   \n",
      "Actinobacteria.Coriobacteriia.Coriobacteriales....          0.033326   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...          0.038161   \n",
      "\n",
      "                                                    mean rel abund  index  \n",
      "Full Taxa                                                                  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...        0.020568      1  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...        0.010144      2  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...        0.051277      2  \n",
      "Actinobacteria.Coriobacteriia.Coriobacteriales....        0.018011      2  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...        0.020624      3  \n",
      "                                                    value percentage  \\\n",
      "Full Taxa                                                              \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...          0.198031   \n",
      "Actinobacteria.Coriobacteriia.Coriobacteriales....          0.025689   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Bactero...          0.017479   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Porphyr...          0.075268   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Prevote...          0.247336   \n",
      "\n",
      "                                                    mean rel abund  index  \n",
      "Full Taxa                                                                  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...        0.064108      2  \n",
      "Actinobacteria.Coriobacteriia.Coriobacteriales....        0.008316      1  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Bactero...        0.005658      1  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Porphyr...        0.024366      2  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Prevote...        0.080069      4  \n",
      "                                                    value percentage  \\\n",
      "Full Taxa                                                              \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...          0.019013   \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...          0.016536   \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...          0.014857   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Bactero...          0.201648   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...          0.095844   \n",
      "\n",
      "                                                    mean rel abund  index  \n",
      "Full Taxa                                                                  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...        0.006755      1  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...        0.005875      5  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...        0.005278      2  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Bactero...        0.071640      4  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...        0.034051     13  \n",
      "                                                    value percentage  \\\n",
      "Full Taxa                                                              \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...          0.014506   \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...          0.032458   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...          0.065488   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Prevote...          0.397060   \n",
      "Firmicutes.Bacilli.Bacillales.Staphylococcaceae...          0.932694   \n",
      "\n",
      "                                                    mean rel abund  index  \n",
      "Full Taxa                                                                  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...        0.005644      1  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...        0.012628      3  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...        0.025479      2  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Prevote...        0.154485      3  \n",
      "Firmicutes.Bacilli.Bacillales.Staphylococcaceae...        0.362886      1  \n",
      "                                                    value percentage  \\\n",
      "Full Taxa                                                              \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...          0.023162   \n",
      "Actinobacteria.Coriobacteriia.Coriobacteriales....          0.019653   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...          0.024830   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Porphyr...          0.049178   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Prevote...          0.048582   \n",
      "\n",
      "                                                    mean rel abund  index  \n",
      "Full Taxa                                                                  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...        0.019925      2  \n",
      "Actinobacteria.Coriobacteriia.Coriobacteriales....        0.016907      1  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...        0.021360      1  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Porphyr...        0.042306      2  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Prevote...        0.041794      2  \n",
      "                                                    value percentage  \\\n",
      "Full Taxa                                                              \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...          0.042543   \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...          0.054168   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Porphyr...          0.100459   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Prevote...          0.279272   \n",
      "Bacteroidetes.Flavobacteriia.Flavobacteriales.F...          0.129655   \n",
      "\n",
      "                                                    mean rel abund  index  \n",
      "Full Taxa                                                                  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...        0.023618      1  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...        0.030072      3  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Porphyr...        0.055771      3  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Prevote...        0.155041      4  \n",
      "Bacteroidetes.Flavobacteriia.Flavobacteriales.F...        0.071979      1  \n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                                    value percentage  \\\n",
      "Full Taxa                                                              \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...          0.020003   \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...          0.017179   \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...          0.024315   \n",
      "Actinobacteria.Coriobacteriia.Coriobacteriales....          0.016111   \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...          0.070213   \n",
      "\n",
      "                                                    mean rel abund  index  \n",
      "Full Taxa                                                                  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...        0.014228      2  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.A...        0.012219      2  \n",
      "Actinobacteria.Actinobacteria.Actinomycetales.M...        0.017296     14  \n",
      "Actinobacteria.Coriobacteriia.Coriobacteriales....        0.011460      3  \n",
      "Bacteroidetes.Bacteroidia.Bacteroidales.Parapre...        0.049943     14  \n",
      "3\n",
      "96\n"
     ]
    }
   ],
   "source": [
    "numbers = [0,1,2]\n",
    "\n",
    "#GF_bile_treshold\n",
    "BA_thresholds_dict = {'GF':202}\n",
    "\n",
    "BAL_GF_ave_groups_to_plot = []\n",
    "BAL_GF_ave_groups_to_plot_with_index = []\n",
    "BAL_GF_ave_groups_to_calculate = []\n",
    "\n",
    "\n",
    "BA = 'Total Bile Acid GF'\n",
    "pick = 'GF'\n",
    "    \n",
    "BAplot_df = []\n",
    "BAplot_df_with_index = []\n",
    "BA_LTother_df = []\n",
    "\n",
    "for number in numbers:\n",
    "    # Get the SamID in each group\n",
    "    samID = taxa_BLAST_bar_plotting_ALL[number]['index'].tolist()\n",
    "\n",
    "    # Select those samID from the metadata\n",
    "    samID_metadata = shannon_meta_initial[shannon_meta_initial['SamID'].isin(samID)]\n",
    "\n",
    "    # Select high/low bile acid, LT vs non-LT\n",
    "    high_samID_all = samID_metadata[samID_metadata[BA] >= BA_thresholds_dict[pick]]\n",
    "    low_samID_all = samID_metadata[samID_metadata[BA] < BA_thresholds_dict[pick]]\n",
    "    samID_all_list = [high_samID_all, low_samID_all] # high, low\n",
    "\n",
    "    # Separate on Lt vs non-LT\n",
    "    sep = ['Lung Transplant', 'Other']\n",
    "    high_low_group = []\n",
    "    high_low_group_with_index = []\n",
    "    isep_relabund = []\n",
    "\n",
    "    for i, group in enumerate(samID_all_list):\n",
    "\n",
    "        isep_group = []\n",
    "        isep_group_with_index = []\n",
    "\n",
    "        for isep in sep:\n",
    "            samID_group = group[group['LT_plot'] == isep]['SamID'].tolist()\n",
    "            # Select the high/low samID from the rel.abund. df\n",
    "            relabund = taxa_BLAST_bar_plotting_ALL[number][taxa_BLAST_bar_plotting_ALL[number]['index'].isin(samID_group)]\n",
    "\n",
    "            # How many patients share each taxa\n",
    "            shared_taxa = relabund.groupby(['Full Taxa']).count()[['index']]\n",
    "\n",
    "            # Mean of taxa\n",
    "            mean_df = relabund.groupby('Full Taxa').mean()\n",
    "\n",
    "            # Take the sum of the mean rel. abund. in high/low group\n",
    "            summie = sum(relabund.groupby('Full Taxa').mean()['value percentage'].tolist())\n",
    "\n",
    "            # Then make the relativ abundance of the mean group -->  this is for plotting\n",
    "            mean_df['mean rel abund'] = mean_df['value percentage']/summie\n",
    "\n",
    "             # Remove taxa lower than limit \n",
    "            limit = 0.0001\n",
    "            mean_df_limit = mean_df[mean_df['mean rel abund'] > limit]\n",
    "\n",
    "            # Increase 'Other' with those relative abundances that were removed bc they were too small\n",
    "            add_other = 1 - mean_df_limit['mean rel abund'].sum()\n",
    "\n",
    "            if 'Other' in mean_df_limit.index:\n",
    "                mean_df_limit.at['Other', 'mean rel abund'] = mean_df_limit.loc['Other', 'mean rel abund'] + add_other\n",
    "            else:\n",
    "                mean_df_limit.loc['Other'] = [0] + [add_other]  \n",
    "\n",
    "            # Merge mean value percentage wtih the number of occurances in patients\n",
    "            finalDF = mean_df_limit.merge(shared_taxa, left_index=True, right_index=True)\n",
    "\n",
    "            print(finalDF.head())\n",
    "            # Concat hig and low\n",
    "            if i == 0:\n",
    "                first = 'high'\n",
    "                second = isep\n",
    "                third = str(len(list(set(samID_group))))\n",
    "                finalDF['type'] = first + ' - ' + second + ' (' + third + ')'\n",
    "                relabund['type'] = first + ' - ' + second + ' (' + third + ')'\n",
    "            elif i == 1:\n",
    "                first = 'low'\n",
    "                second = isep\n",
    "                third = str(len(list(set(samID_group))))\n",
    "                finalDF['type'] = first + ' - ' + second + ' (' + third + ')'\n",
    "                relabund['type'] = first + ' - ' + second + ' (' + third + ')'\n",
    "\n",
    "            isep_group_with_index.append(finalDF[['mean rel abund', 'type', 'index']])\n",
    "            isep_group.append(finalDF[['mean rel abund', 'type']]) # LT vs non-LT\n",
    "            isep_relabund.append(relabund)\n",
    "\n",
    "        df_limit_with_index = pd.concat(isep_group_with_index)\n",
    "        df_limit = pd.concat(isep_group) # Combining LT vs non-LT\n",
    "        df_limit_with_index.reset_index(inplace=True)\n",
    "        df_limit.reset_index(inplace=True)\n",
    "        high_low_group_with_index.append(df_limit_with_index)\n",
    "        high_low_group.append(df_limit) # high vs low\n",
    "\n",
    "    hl_group = pd.concat(high_low_group) # combining high/low\n",
    "    hl_group_with_index = pd.concat(high_low_group_with_index)\n",
    "\n",
    "    BAplot_df.append(hl_group)\n",
    "    BAplot_df_with_index.append(hl_group_with_index)\n",
    "    BA_LTother_df.append(isep_relabund)\n",
    "\n",
    "print(len(BAplot_df))\n",
    "print(len(BAplot_df[0]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "def order_list(mylist):\n",
    "    \"\"\"\n",
    "    Order list based on ['high - Lung Transplant','high - Other', 'low - Other','low - Lung Transplant']\n",
    "    \"\"\"\n",
    "    for item in mylist:\n",
    "        if item.startswith('high - Lung Transplant'):\n",
    "            a = item\n",
    "        elif item.startswith('low - Lung Transplant'):\n",
    "            b = item\n",
    "        elif item.startswith('high - Other'):\n",
    "            c = item\n",
    "        elif item.startswith('low - Other'):\n",
    "            d = item\n",
    "    return [a, b, c, d]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Text(0.5, 1.0, 'Oropharynx')"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAABLMAAANcCAYAAACkN+27AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMiwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8vihELAAAACXBIWXMAAAsTAAALEwEAmpwYAAA550lEQVR4nO3de7zldV0v/tdnGO8ixiVRnJEckkZSjzIh3pixPBMYZp34RRcqcxAtrYcm+UOUnFDK8JJ46RgNecESOKZmnDKVfpOGEo79NEXL8MIBPShoXuEY43zOH3tp4zCz99p7r+981mft5/Px4LH32nvt73rN97HW/jCveX++q9RaAwAAAAA9WNU6AAAAAACMS5kFAAAAQDeUWQAAAAB0Q5kFAAAAQDeUWQAAAAB0Q5kFALAXpZQ7l1IObp0DAIDvpcxixSqlbC2lvL2UcnEp5Y9GX/vjUsqzR59vKqWc3jYlAJNQSrlDKeWVpZTXjX7v//wYP3ZqkoftdozDSylPmecxLi2lPLWU8qa9fO+kUsqG3W5vLaUctdg/BwDDGa0VryqlbCulvLWU8mOL+Nnb/e4HhrO6dQBo7Mxa67WjUuuQ0dceU0p5edNUAEzaU5L8da31nUlSSnlcKaUk+e+j738jye8neVWS25K8Psmjk3yrlPLoJEcm+dMk9yulHJ9kS5J7Jnl6rfWLpZSHJ3lgkldnbh15UpIbaq3vKaW8Psn2JLeVUp6R5Kgkj0ziLz4A02VLkstrrX9bSlmd5C9LKa9OcnmS9yc5Ocmdkjwrc+vHh5MckuS5SY4c/QP5vZL8bJIXJTkoyd1qrb9SSvlokncmWVNr/blSyvOS/Fnm1pbtSe6Y5JIkZyR5X5J71lr/ZL/8qaFDJrNY6f6glPL2JDuT/GLmFqodSTa3DAXAxD0kyT+WUg4dlUtnZ+4f9d6c5F1Jjk1yj9F/lyf5WJIrk7x99POvSPK/Rp+fnuTXkvxmkl1JUmv9xyQfqbW+b4Ecx9Van5nk/5vAnwmAyXpI5kqr1Fp3JvlEks/VWn87ySm11i1JXpnktCTfqrW+KHNF1H9N8uVa66+PfuY+Sf569N8DR8f+1Og47y+lHJfkfqOvf7jWem6SI2utH0vylSQ/o8iC+SmzWOn+31rrTyX5t8z9ReWJmVtwntYwEwCTd02Sx9Vab661PinJF5LcN3PF1HVJ/j3J1zNXch00+lh3+/lv7Pb5HUbfu3vm/kV+b3YmWV1KueceX981+vjtpf5BABjMx5IcnySj6d37J7lx9L2y2/1W7Xb7zpn7nf+10e2dmfuHkWcl+VKSz4++/p115OLMTQK/d3T7lt1+LkkOztz6ccfl/3FgdimzWOleWkp5Y5IfTnJBrfX0WusvZO4vKWuTbCmlvKmU8tKmKQFYrtcmeWwp5X+UUt6S5KOZ+wvEPZP8cpJ7Z+7/i16Q5LjM/cv8dZkru/b0p0kuTPLCJF/ex+O9P3OTW89L8s3dvn7VaCv7Ccv88wAwef89yY+PJngvTfKa3b73llLKH2duG+AbkhxYSvnDJJuS/N0ex9mZ5IAk/0+S++7+Dxu11n/P3PTVW/d88FLKjya5IXP/yH72BP48MLNKrXXhewEAAABJ5i74Xms9bQk/95wktdb6kgFiwYqhzAIAAACgG7YZAgAAANANZRYAAAAA3VBmAQAAANCN1a0DJMmJJ55Y3/nOd7aOATCNysJ3mX3WCYB9sk7EOgEwj5lcJ6ZiMuvmm29uHQGAKWadAGA+1gmAlWUqyiwAAAAAGIcyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBurJ7UgUopd0zyoiTvrbVeXko5Ocm6JHeqtZ4/qccBoE/WCQDmY50AYFwTK7OSHJ7kI7vd3lRrPbOUcl4pZVWtddfudy6lnJHkjCRZu3btxEKUUiZ2rGlTax30+M+58qxBj9/K+Y968aDHP/Ge5w16/Fbe+ZXnDXr8Pz1906DHb+nJ27a3jjCtpmKd2PADh03sWNNmx2duGvT4s/q6Hfo1+7yffNigx2/lvHf806DH37Bhw6DHb2nHjh2tI0wr68TAhlwn/vKQwwc7dmtP/NKNwz7Aj79r2OO38rebBz38rK4T1ojxTGybYa31fyX53G5f2jn6eMDeHqfWemGtdUOtdcNhh83uggHAHOsEAPOxTgAwriGumXVAKeXxSa4upTwryc211p0L/RAAK4Z1AoD5WCcAmNcktxmm1rp9kscDYLZYJwCYj3UCgHF4N0MAAAAAuqHMAgAAAKAbE91mCDCuT33xa60jAAAA0CGTWQAAAAB0w2QWAABT56ee8OLWEYBFeOI/HNw6AivMj99nV+sINGQyCwAAAIBuKLMAAAAA6MbMbTP82Yevax0BAAAAgIHMXJkFAAAAvfjIUz7ZOsIgHpLNrSMww5RZAAAALM/1a1onGM761gGAPblmFgAAAADdMJkFAAAAdOU177mmdYRBnNc6QCdMZgEAAADQDZNZAAAD8i/HAACTZTILAAAAgG4oswAAAADohm2GAAAA0Mg7LtneOsIgHnLKM1pHYIaZzAIAAACgGyaz+K6PfPPFrSMAAAAAzMtkFgAAAADdUGYBAAAA0A1lFgAAAADdcM0sAAAAoCs3P/m/to5AQ8osWKZfO+BVrSMM5HmtAwAAAMDtKLMAAJg6RxxxYOsIAPvFm0+4tXWEQZzTOgAzzTWzAAAAAOiGMgsAAACAbiizAAAAAOiGa2bBMr3wB+7bOsIgntg6AAAAAOyFySwAAAAAumEyi+86+05/3zrCQDa2DgAwE/72Yze0jjCIJw98/E9+5uaBHwEAYGUxmQUAAABAN0xm8V2PuenPW0cYiMksAAAAmBXKLABgLL9z0ZWtIwAAgG2GAAAAAPTDZBYAAMDIFUce0ToCAAswmQUAAABAN5RZAAAAAHTDNkNYpisO3Nk6AgAAAKwYJrMAAAAA6IYyCwAAAIBu2GYIAAAAjXziyY9uHQG6YzILAAAAgG4oswAAAADohjILAAAAgG4oswAAAADohjILAAAAgG4oswAAAADohjILAAAAgG6sbh0AAAAAVqqPH/fc1hGG8fHWAZhlyiwAAACW5euXf6R1hMEcuLl1AmBPthkCAAAA0A1lFgAAAADdUGYBAAAA0A1lFgAAAADdmLkLwH/qC19tHQEAAACAgZjMAgAAAKAbyiwAAAAAujFz2wwBAKbJwS/8+dYRhvHqy1snAABWKJNZAAAAAHRj5iaz3nD5x1pHAAAAAGAgJrMAAAAA6MbMTWYBAEyT46+6sXWEQexoHQBgRnzyC4e3jjCI9ZnN9Y/poMwCAABgWe7x+bu2jjCY2joAcDu2GQIAAADQjZmbzDrs0Nn9FwEAAACAlc5kFgAAAADdmLnJLACAaXLRWU9qHQEAYKYos2CZfuzrs/kyGvpdqu5y7PkDPwIAAACzaDb/Fg4AAADMrAM/8rDWEQZxa+sAnVBmAQAAQCNP/IeDW0eA7iizAAAGtOXFr28dYRA7TnlG6wgAwAqlzAIAABj58NYLWkcYzMbWAQAmZFXrAAAAAAAwLpNZAAAAQFdu/Zs/bB1hIOe2DtAFk1kAAAAAdMNkFgAAAMuy+RddkQvYf0xmAQAAANANk1kAAAP64F+8vHUEYBE2Xntx6wjD2Wh6CpgNJrMAAAAA6IYyCwAAAIBuKLMAAAAA6IZrZgEAAABd2bDp6NYRBrFjR+sEfVBmwTLt+L1DWkcAAACAFUOZBcCKcs7XDmgdAQDgP12/pnWCYaxvHYBZ5ppZAAAAAHRDmQUAAABAN5RZAAAAAHRDmQUAAABAN1wAHgAAgGV5191+rHUEYAUxmQUAAABAN5RZAAAAAHTDNkMAVpRNDz68dQQAAGAZTGYBAAAA0A1lFgAAAADdUGYBAAAA0A3XzIJl2nXNra0jDGLV5tYJAAAA4PaUWUATRxxxYOsIAAAAdMg2QwAAAAC6YTILAADgOz7wiNYJhrOldQCAyTCZBQAAAEA3TGYBAABAI+WyF7eOMIg68BtKXXXV9mEfgKlmMgsAAACAbpjMgmU67s9uaR1hEDue1ToBAAAA3J7JLAAAAAC6ocwCAAAAoBvKLAAAAAC64ZpZAAAAQFeOP35T6wiD2LFjR+sIXVBmwTJ96OSXtY4AAAAAK4ZthgAAAAB0Q5kFAAAAQDeUWQAAAAB0Q5kFAAAAQDeUWQAAAAB0w7sZAgAwdY5Ye3DrCABMsVsu3tE6Ag2ZzAIAAACgGyazAAAGtOqQK1tHGETNCa0jAAArlMksAAAAALqhzAIAAACgG8osAAAAALqhzAIAAACgG8osAAAAALqhzAIAAACgG8osAAAAALqxunWASfvEv/xT6wiD+f57bWwdAQAAAJp78zWvbh1hGOuf0TpBF0xmAQAAANANZRYAAAAA3Zi5bYYAAPTv7LN/vnWEwWzevKN1BADomsksAAAAALqhzAIAAACgGzO3zXDjxde2jjCcjd7NEAAAAFjZTGYBAAAA0A1lFgAAAADdUGYBAAAA0I2Zu2YWAAD9u/qNF7SOALBfHHvF5tYRBnLToEd/xyXbBz1+Kw855RmtI3Rh5sqsXSde3TrCYFZlS+sIAAAAAE3ZZggAAABAN5RZAAAAAHRj5rYZAgDQv7s99W9bRxjMre97VOsIANA1k1kAAAAAdMNkFgAAU+fWv/nD1hEGdG7rAADQNZNZAAAAAHTDZBYAAADQlTefcGvrCIM4p3WATiizAAAG9PHjnts6wjA+3joAALBS2WYIAAAAQDdMZgEAALA8lx3TOsFwNrcOAOzJZBYAAAAA3VBmAQAAANAN2wwBAJg6GzYd3TrCYHbsaJ0AAPo2e2XW2oe2TgAAAADAQCZWZpVSTk6yLsmdaq3nl1KenOSeo9u/P6nHAaBP1gkA5mOdAGBck5zM2lRrPbOUcl4pZVWSkuSwJN/e251LKWckOSNJ1q5dO8EYQA+OWHtw6wjsf9YJAOZjnQBgLJMss3aOPh6QuQvL37/W+txSyitKKatqrbt2v3Ot9cIkFybJhg0b6gRzADCdrBMAzGc61olLT53YoabOttYBACZjkmXW1aWUZyW5NcnmJDeVUn4rydf3XHgAWJGsEwDMxzoBjO0TT3506wg0NLEyq9b61j2+9NeTOjYA/bNO9O+YVY9tHWEgH28dAIh1AoDxrWodAAAAAADGpcwCAAAAoBuTvGYWADDLrl/TOsEw1g97+J/5i2GP38rQmzOfeOSRAz8C7N2ucx/fOsJgVuW9rSMATITJLAAAAAC6ocwCAAAAoBu2GQIAAABd+fhxz20dYRjeZHksJrMAAAAA6IbJLKCJs8/++dYRBrN5847WEQC695ef/WzrCIM5p3UAAOicySwAAAAAuqHMAgAAAKAbthkCAAzo+jWtEwAAzBaTWQAAAAB0Q5kFAAAAQDeUWQAAAAB0wzWzAAAAgK588guHt44wiPW5sXWELpjMAgAAAKAbyiwAAAAAuqHMAgAAAKAbrpkFAMDU+dAhl7aOAABMKZNZAAAAAHRDmQUAAABAN2wzBADGsuHsL7WOMIgdm1snAABgMZRZQBM7tv9r6wgAAAB0yDZDAAAAALphMgsAGMuOk5/QOgIAACizAAAAgL488R8Obh2BhpRZAAADevzjjmodYRhXXds6AQCwQimzgCY2bDq6dYTB7NjROgEAAMDscgF4AAAAALqhzAIAAACgG8osAAAAALrhmllAEzsuvqV1BACm2Aseua51BABgSpnMAgAAAKAbJrMAAABYlmOv2Nw6woBuah0A2IMyCwAYy4bL/6p1hEHs2PqC1hEAAFgEZRYs1w33aJ2gT5ee2jrBcLa2DgDQv9s+9LDWEQb0T60DAFPkiiOPaB0BuuOaWQAAAAB0Q5kFAAAAQDeUWQAAAAB0Q5kFAAAAQDeUWQAAAAB0Q5kFAAAAQDeUWQAAAAB0Q5kFAAAAQDeUWQAAAAB0Y3XrANC9969pnQAAAGBluX5G/x62vnWAPiizAACYOq95zzWtIwzmvNYBAKBzthkCAAAA0A2TWQAAANDIh7de0DrCIDa2DsBMM5kFAAAAQDeUWQAAAAB0wzZDAACA71jjrcQApp0yC2jjA49onYAValavS5G4NgUAACuDbYYAAAAAdEOZBQAAAEA3lFkAAAAAdEOZBQAAAEA3lFkAAAAAdMO7GcIyHXvXE1tHGMiOQY++4UtnD3r8lnZkc+sIAAAAM8tkFgAAAADdMJkFAMDU+eRnbm4dAQCYUsosoImrrtreOgIAAAAdss0QAAAAgG4oswAAAADohm2GsEwnP2HYd/0DAAAA/pMyC2jiun96e+sIg1l33GmtIwAAAMws2wwBAAAA6IbJLAAAps7jf+KxrSMMZscOlyhg9lxx5BGtIwAriMksAAAAALphMguW6Yb7tk4AAAAAK4cyCwCAqfPBv3h56wgAwJSauTKr3OuBrSMAAAAAMJCZK7MAYD4br724dYThbNzYOgEAAAzOBeABAAAA6IbJLKCJo87/ndYRBlPfclrrCMAUufR1d2wdAQBgppjMAgAAAKAbyiwAAAAAumGbIQAAADQys29O441pGJAyC5bp/Y9snQBYlA88onWC4WxpHQAAAIZnmyEAAAAA3TCZBQAwpOvXtE4wjPWtAwCwkpX8UusIg6itA3TCZBYAAAAA3TCZBQDA1Fl1yJWtIwym5oTWEQCgayazAAAAAOiGySxYplm9FAoAAABMI2UWAAAAtPKBR7ROMIwtrQMwy2wzBAAAAKAbyiwAAAAAumGbISzTr33krNYRhvGoF7dOAEyZC055YusIAAAwe2VWOeh+rSMAAAAAMJCZK7MAAKZJuWw2J13r5mGPf/c8d9gHAAC65ZpZAAAAAHRDmQUAAABAN2wzhGW67tlvaR1hGFfN5rYYAAAA+qbMAgDGcvWfn9s6wiAeddY5rSMAALAIyiwAAKbOBc/c1DrCcLZtb50AoH+XHdM6wTAGfoOVWaHMAgAY0LFXzOr/ld7UOgAAsEK5ADwAAAAA3VBmAQAAANAN2wwBAJg6f/uxG1pHGMyTWwcAgM6ZzAIAAACgG8osAAAAALqhzAIAAACgG8osAAAAALqhzAIAAACgG8osAAAAALqhzAIAAACgG8osAAAAALqhzAIAAACgG8osAAAAALqhzAIAAACgG8osAAAAALqhzAIAAACgG8osAAAAALqhzAIAAACgG8osAAAAALqhzAIAAACgG6tbBwAAgD39zkVXto4AAEwpk1kAAAAAdEOZBQAAAEA3bDMEAAAAunLsFZtbRxjITa0DdGHmyqyLbvyb1hEGs+WeT2sdAQAAAKAp2wwBAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6Mbq1gEAAGBPx6x6bOsIA/p46wAA0DWTWQAAAAB0w2QWAAAAy/LhrRe0jjCYja0DALdjMgsAAACAbiizAAAAAOiGMgsAAACAbiizAAAAAOiGMgsAAACAbiizAAAAAOjG6tYBAABm2RVHHtE6AgDATDGZBQAAAEA3lFkAAAAAdEOZBQAAAEA3lFkAAAAAdEOZBQAAAEA3vJshAADAyIEf/FDrCIP55inDHXvjxdcOd/DWNm5snQDYg8ksAAAAALqhzAIAAACgG8osAAAAALqhzAIAAACgG8osAAAAALqhzAIAAACgG6sndaBSyslJ1iW5U631/FLKTyRZn+QutdYXTupxAOiTdQKA+VgnABjXxMqsJJtqrWeWUs4rpaxKcmKSG5J8bm93LqWckeSMJFm7du0EYwAwpawTnTv6HrtaRwBmm3UCgLFMsszaOfp4QOa2L35/rfU3SimvKqWsqrV+z/8B11ovTHJhkmzYsKFOMAcA08k6AYzv+jWtEwxnfesAU8s6AcBYJnnNrKtLKc9KcmuSzUneVko5M8mX9lx4AFiRrBMAzMc6AcBYJjaZVWt966SOBcDssU4AMB/rBADj8m6GAAAAAHRDmQUAAABAN5RZAAAAAHRDmQUAAABAN5RZAAAAAHRDmQUAAABAN5RZAAAAAHRjdesAAAAA0+LpP/m41hEAWIAyCwAAAFq59NTWCYaxrXUAZplthgAAAAB0w2QWADCWxzzssNYRAADAZBYAAAAA/VBmAQAAANANZRYAAAAA3XDNLAAAps/7H9k6wXA2tw4AAH0zmQUAAABAN0xmAQBjOfwu920dYRDfbB0AAIBFMZkFAAAAQDfmncwqpdw5ydOS/EiS/0jyoSSvrbXu3A/ZAAAAAOB7LLTN8NeSvKPW+ookKaU8IMmvJ3nlwLmAGbf+hPWtIwAAANChecusWusfJkkpZVWShyX5ZK1VkQUAAABAE/NeM6uUcsfRp89PcnCSFw6eCAAAAAD2YaFthr9dSvlUkmuTHJ/kxuEjAQAAAMDeLbTN8LxSyg8l+cUk76u1vmv/xAIAAACA21tom+HLkjyu1npOkltLKeftn1gAAAAAcHsLbTO8Q5K7J0mt9X1J3jd4IgAAAADYh4XKrD9I8r93/0Ip5aBa61eHiwSsBHd94xdaRxjOb7YOANC/DZf/VesIg9mx9QWtIzCPow85snUEYAxXHHlE6wg0tFCZ9VNJvq+Uck2SnUk2JLk+ybaBcwEAAADA7Sx0AfjXlFIOTfKY0X1fXWu9ab8kA2bah7b/TOsIAPvFh7de0DrCIDa2DgAArFgLTWal1npzkrfthywAAAAAMK8FyywAANjffu1Np7eOAABMqQXLrFLKs5P8SJK3J7lPrfXlQ4cCAAAAgL1ZNcZ9jkzy6VrrJUmOHjYOAAAAAOzbOGXWt5P8YCnlqUkOHDgPAAAAAOzTONfMOifJY5MclOSpw8YBAAAAgH0bZzLrtUn+Ock1SV49bBwAAAAA2LdxJrO+XGv9bJLPllK+OXAeYKV4/5rWCYazuXUAgP4dddD9W0cAAKbUOGXWFaWU1yU5IMlfDZwHAAAAAPZpnDLrnzN34fdVSe4ybBwAAAAA2LdxyqytSd6QZOewUQAAAGBl2XXu41tHGMSqvLd1BGbYOGXWtUn+McktA2cBVpKXPaJ1guG4ZhYAAMBgximz7pPklbvdfvJAWQAAAABgXuOUWRck+ZEkJclBw8YBAAAAgH0bp8x6cZJ/S/LRzJVaAACMaeO1F7eOMIyNG1snAABWqFVj3OemJO9I8tUkXx42DgAAAADs2zhl1iVJbkzyo6OPAAAAANDEvNsMSykvTFJHN7+a5C6DJwIAAACAfVjomlnbkpyT5CVJdiV55tCBAABg0zmHtY4wmLqtdQIA6Nu8ZVat9bpSyh0zd62skuSu+yUVAAAAAOzFOO9m+MIkZ2Zuu+ELh40DAAAAAPs2zgXgH5Tk0CT3SvL8YeMAAAAAwL6NM5l1QpIzaq3fHjoMAAAAAMxnnDLrrklOKqXckiS11r8bNhKwEmxbs651BID94wOPaJ1gGFtaBwAAVqpxyqwrkxw8+g8AAAAAmhmnzNq02+ffSPLGYaIAAAAAwPwWLLNqrb+aJKWUg5KcPXgiAAAAANiHBcusUsqPjj69Q5IfHjYOAAAAAOzbONsM1ySpSXYmecawcQAAAGAFWbO+dQLozjhl1v9J8lOjj7cm+cyQgYCV4dC3Hd46wnC23dg6AQAAwMwap8w6odb680lSStma5G2DJgIAYMU79orNrSMM6KbWAQCga/OWWaWULUl+qJTy1CTfSvLQ/ZIKAAAAAPZiocmsa5Ocm7lrZiXJy4aNA/351Be+2jpClzY9eIa3GQIAADCYecusWuvfl1Kek+T4JLcluUuS9+6PYAAAAACwp3GumfWDSS5J8vEkTx02DgAAAADs2zhl1h2S3Jjk9CT3GTYOAAzs0lNbJxjOttYBAABgeOOUWb+epCT5dpKvDJoGAACA7uw65g2tIwxmVba0jgDsYcEyq9Z6y+jTKwfOAgAAAADzWjXOnUopB5ZSxrovAAAAAAxlwcmsUsrvJ7l/kreUUo6ttZ41fCwAAAAAuL1xrpl1tySfqrX+j1LKjw4dCACYTo8+/kGtIwAAMJBSymeT3K91jgVcV2s9cpwy62tJHj6a0LLVEICufeqKt7eOMJh1Oa11BAAA+nW/WmtpHWI+pZSajDeZ9a4kf57kgFrrRwdNBQAAALCAD2+9oHWEQWxsHaAT45RZ65L8QpIvl1LuXGv94MCZAAAAAGCvFtw2WGt9XZLXZe7aWWcPnggAAACAqVdKOaLF447zboZ/muTKJC+otX5l8EQAAAAANFdK2ZrkIUmOSvKRJGfVWm/Y7S4vTvJLe/zMpiRH1Vq3jW6/PskdMjdQ9S+11t8tpZyT5PokN9Ra37PYXPOWWaWU05N8LsmRSZ5dSqm11t9Z7IMAAAAA0Jda69ZSypFJTq+1Pr+U8qpSSpJ8OXPXWP8vpZRjkjwnyW1J/i3JP+7lUL9Sa91ZSvmr0e2jM1dmpZTypCQPS3JgkmfWWr+6UK6FJrPekeQuSWqSkuTuCx0QAJhNP7v+4a0jAADQyKi0+nyt9fdLKa9J8onMTWtdl7nLUx2e5Oey9zLrolLK4Un+ZS/fe0qS9yW5a5KHJtm+UJZ9llmllPtkblzsIaNwJcm9kmxe6KAAAAAAzJRVmRt2+o4y+vhfkvx0km1J/mMfP7tlNJn15lHftLuv1VrPKqUcn+SL4wTZZ5lVa/18kl8upTyy1vr+JCmjWTIAAAAAVo5a60dLKU8ppbw0yf+utX5pNG311SRrkzwtc9NVe/OGUsquzBViX9jje/+zlHJR5nYD/uo4WRa8AHySny6lvDDJvye5d5JHjXNgAGC2HHXQ/VtHAABgP6u1fjbJ80ef/+Ye33vc6NOf3uPHtu92nyft5bCn7fb5ZC8AP3L3JK/O3B7IX17sAwAAAADApKwa835fSvLMJD8wXBQAAAAApkkpZXMp5f17+fpJpZQNu93eWko5arfb5yxw3N8opbyxlHJRKeX3Rl97fSnlz0opbyql/Pa+fnbeyaxSyqokzxjd/Hzm9kECAAAAsDL8QpJ3llJOSHJCksOS3Jq5dya8rZTyjCRHJXlkkjft9nNHl1I2JfmNJP+auetsvSr57jsj3qfW+suj248updx59HO/UmvdOV+ghbYZXpTvvVJ9kjx5gZ8BgKn1A1de2DrCcI47beH7AADAmEophyS5Q5LXJXllks/XWp9eSnlQkmNHdzuu1vrLpZQ/2Mdh/r7W+spSypuSvGr0tQcnuXr0GC9N8oAk546+d1EppSb541rrB/Z2wHnLrFrrr5ZSjk3ymCQfSXLjGH9WAAAAAPp3WuYmsZ6XZFOSS0Zfv99u99k1+vjtfRzjltHH3aetrkny1CRvq7WeWUp5auau2Z4kW5Y7mZUkT8/cNbP+PMkfJTlljJ8BAGbMpo9+sXWEQdR7t04AADC1Tkxycq31P0op70pyaSllZ5KvJPnU6D5XlVJenuS4JH86zkFrrf9cSvlkKeVtSf5P5i5rdUmSJ43z8+OUWbckuW30cd5mDAAAAIDZUGs9abfP35q5LYd7es8+fvY718DYPrr9pD2+f0GSC/b4sSdlDOO8m+HFSdYkuXAvDwIAAAAA+81C72b4gCTHJ/nNzO19PC/JXi++BQAAAABDW2ib4YuTnJ+5a2XdI8mLBk8ErAg/9vVxdjn3aUfrAAAAADNsob9N3lhrvaqU8qIkp9Rav7IfMgEAAACwf11XSqmtQyzgumThMusepZT7Z+4q9QeXUg6utX566GQAAAB0ZM361gmAZaq1Htk6w7gWKrP+LckvJfno6GNNcu7QoYDZt+P3DmkdAQAAgA7NW2bVWn93fwUBAAAAgIXM7hWYAYDJuuyY1gmGsbl1AAAAFmNV6wAAAAAAMC5lFgAAAADdUGYBAAAA0A1lFgAAAADdUGYBAAAA0A1lFgAAAADdUGYBAAAA0I3VrQMAAAAALMbGay9uHWEYGze2TtAFk1kAAAAAdMNkFtDErmtubR1hMKs2t04AAAAwu5RZAKwsa9a3TgAAACyDbYYAAAAAdMNkFtDEAV99YesIg6mtAwAAAMwwZRYs08te/5bWEQAAAGDFsM0QAAAAgG4oswAAAADohjILAAAAgG4oswAAAADohgvAAwBjOfbDT2kdYSA7Bj36rmPeMOjxW1mVLa0jAAArlMksAAAAALphMgsAAADoywce0TrBMAw+j0WZBQDA1Dnnawe0jgAATCnbDAEAAADohjILAAAAgG4oswAAAADohmtmAQDj+dJ1rROwgmx68OGtIwAAU8pkFgAAAADdUGYBAAAA0A1lFgAAAADdUGYBAAAA0A0XgAfauOEerRMAAADQIZNZAAAAAHRDmQUAAABAN5RZAAAAAHRDmQUAAABAN5RZAAAAAHTDuxkCbbx/TesEAAAAdMhkFgAAAADdMJkFy7Tx4mtbRxjGxo2tEwDMhjXrWycAAJgpJrMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6Mbq1gGAlekFpx7aOgIAAAAdMpkFAAAAQDeUWQAAAAB0Q5kFAAAAQDeUWQAAAAB0Q5kFAAAAQDeUWQAAAAB0Y/WkDlRKOTnJuiR3qrWeP/raE5I8ptb6nEk9DgB9sk4AMB/rBADjmliZlWRTrfXMUsp5pZRVSR6e5LbsY/qrlHJGkjOSZO3atROMAfvXrhOvbh1hEKuypXUEZo91AoD5WCcAGMsky6ydo48HZG7BeUKSLyZ5WCnlHrXWr+1+51rrhUkuTJINGzbUCeYAYDpZJ4CxfXjrBa0jDGZj6wDTyzoBwFgmWWZdXUp5VpJbk2yutZ6dJKWU++658ACwIlknAJiPdQKAsUyszKq1vnUfXz9zUo8BQL+sEwDMxzoBwLi8myEAAAAA3ZjkNkMAYIad87UDWkcAAACTWQAAAAD0Q5kFAAAAQDdsMwQAxrLpwYe3jgAAACazAAAAAOiHMgsAAACAbiizAAAAAOiGMgsAAACAbiizAAAAAOiGdzMEYGVZ+9DWCQAAgGUwmQUAAABAN5RZAAAAAHRDmQUAAABAN1wzC2jihvu2TsBKteoBJ7WOAIxh47UXt44wnI0bWycAgK6ZzAIAAACgG8osAAAAALqhzAIAAACgG66ZBQAAAPTl0lNbJxjGttYB+qDMApo4+IfOah1hQC9uHQCgfx94ROsEw9nSOgAA9E2ZBcCKUk5/bOsIg6lv+XTrCAAAMDjXzAIAAACgGyazAICxfHjrBa0jDGJj6wAAACyKySwAAAAAuqHMAgAAAKAbthkCTbzk0X/QOsJgzq/ezRAAAGAoJrMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuuAA8AADT59JTWycYzrbWAQCgbyazAAAAAOiGMgsAAACAbthmCMCKsv6E9a0jAGP41BVvbx1hMOtyWusIANA1k1kAAAAAdEOZBQAAAEA3bDMEAMay8eJrW0cYxsaNrRMAALAIyiwAAACgK7vOfXzrCINYlfe2jtAF2wwBAAAA6IbJLKCJn334utYRAAAA6JAyCwCAqfMDV17YOsJwjjutdQIA6JpthgAAAAB0w2QWACvKycc+qHUEAABgGZRZAABMnzXrWycAAKaUbYYAAAAAdEOZBQAAAEA3lFkAAAAAdEOZBQAAAEA3lFkAAAAAdEOZBQAAAEA3VrcOAAAAALAoa9a3TkBDJrMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IZ3MwSa+NQXvto6AgAAAB0ymQUAAABAN5RZAAAAAHRDmQUAAABAN5RZAAAAAHRDmQUAAABAN5RZAAAAAHRDmQUAAABAN5RZAAAAAHRDmQUAAABAN1a3DgAAMNPWPrR1AgCAmWIyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6Mbq1gEAAADo24Ef/FDrCIP55imtEwB7MpkFAAAAQDeUWQAAAAB0wzZDAFaUpz74Sa0jAAAAy2AyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYLwAMA47n01NYJhrGtdQAAABbDZBYAAAAA3VBmAQAAANAN2wwBgLHsOvfxrSMMYlXe2zoCAACLoMwCAGD6rH1o6wQAwJSyzRAAAACAbiizAAAAAOiGMgsAAACAbiizAAAAAOiGMgsAAACAbiizAAAAAOiGMgsAAACAbqxuHQAA6MSa9a0TAACAySwAAAAA+mEyC4AV5aiXfaF1hMHUrT/UOgIAAAzOZBYAAAAA3VBmAQAAANAN2wwBAJg6qx5wUusIAMCUMpkFAAAAQDeUWQAAAAB0Q5kFAAAAQDdcMwuWa+1DWyfo0hsu/1jrCAAAAHTIZBYAAAAA3VBmAQAAANANZRYAAAAA3VBmAQAAANANF4AHmjjs0Lu2jgAAAECHTGYBAAAA0A1lFgAAAADdsM0QAAAA6MpZ9/6+1hEGcX7rAJ0wmQUAAABAN5RZAAAAAHRDmQUAAABAN1wzCwCAqVNOf2zrCIOpb/l06wgA0DWTWQAAAAB0Q5kFAAAAQDeUWQAAAAB0wzWzAAAAgK685N0ntY4wiPMf1TpBH0xmAQAAANANZRYAAAAA3VBmAQAAANANZRYAAAAA3VBmAQAAANAN72YINPGJf/mn1hEG8/332tg6AgAAwMwymQUAAABAN5RZAAAAAHTDNkMAAACW5ek/+bjWEYAVxGQWAAAAAN0wmQUAMKBVDzipdQQAgJliMgsAAACAbiizAAAAAOiGbYYAwFgOuOTdrSMMop7SOgFA/44+5MjWEYAVxGQWAAAAAN1QZgEAAADQDWUWAAAAAN1QZgEAAADQDWUWAAAAAN1QZgEAAADQjdWtAwAr08aLr20dYTgbN7ZOAAAAMLNMZgEAAADQDWUWAAAAAN1QZgEAAADQDWUWAAAAAN1QZgEAAADQDe9mCAAAwLKcfsn61hEGs2Vr6wTAnkxmAQAAANANZRYAAAAA3VBmAQAAANANZRYAAAAA3ZjYBeBLKScnWZfkTrXW80spv5JkTZLDa63PmNTjALNh14lXt44wmFXZ0jrCVLJOADAf6wQA45rkZNamWusFSQ4qpaxK8p5a64uS3GF0+3uUUs4opewopey46aabJhgDgCllnQBgPtYJAMYyscmsJDtHHw/IXEn2xVLKuUleW2vdteeda60XJrkwSTZs2FAnmAOA6WSd6Nz6E2b3bdeBqWCdAGAskyyzri6lPCvJrUk2J/nJzC1IP1FK+Vit9bYJPhYA/bFOADCfqVgnTr9kdov7LVtbJwCYjImVWbXWt+7xpb+e1LEB6J91AoD5WCcAGJd3MwQAAACgG8osAAAAALqhzAIAAACgG8osAAAAALoxyXczBABgDw9846+3jjCIj//m/2wdAQBYoUxmAQAAANANZRYAAAAA3VBmAQAAANAN18wCAMZy8rEPah0BAABMZgEAAADQD2UWAAAAAN1QZgEAAADQDWUWAAAAAN1QZgEAAADQDWUWAAAAAN1QZgEAAADQDWUWAAAAAN1QZgEAAADQjdWtAwAAzLKTj31Q6wgAADPFZBYAAAAA3VBmAQAAANAN2wwBAJg6609Y3zoCADClTGYBAAAA0A1lFgAAAADdsM0QAGBAL3n3Sa0jDOL8R7VOAACsVCazAAAAAOiGMgsAAACAbiizAAAAAOiGa2YBADB1Tj72Qa0jAABTymQWAAAAAN1QZgEAAADQDdsMgTbWPrR1AgAAADpkMgsAAACAbiizAAAAAOjGzG0zPOqg+7eOAAAAAMBAZq7MAgCG8ZJ3n9Q6wiDOf1TrBAAALIZthgAAAAB0w2QWAABTZ1YnARPTgACwXCazAAAAAOiGMgsAAACAbiizAAAAAOiGMgsAAACAbiizAAAAAOiGMgsAAACAbqxuHQAAAABWqrPu/X2tIwzi/NYBmGkmswAAAADohsksWKZyrwe2jgAAAAArhsksAAAAALphMgtowkQbAAAAS2EyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6Mbq1gEAYL+67JjWCYaztXUAAAAYnsksAAAAALqhzAIAAACgG7YZAk2Ug+7XOgIAAAAdMpkFAAAAQDeUWQAAAAB0Q5kFAAAAQDeUWQAAAAB0wwXgAQCYPpcd0zrBcLa2DgAAfTOZBQAAAEA3lFkAAAAAdEOZBQAAAEA3lFkAAAAAdEOZBQAAAEA3lFkAAAAAdEOZBQAAAEA3lFkAAAAAdEOZBQAAAEA3lFkAAAAAdGN16wDAynTRjX/TOsJgttzzaa0jAAAAzCyTWQAAAAB0Q5kFAAAAQDeUWQAAAAB0Q5kFAAAAQDeUWQAAAAB0w7sZAk0cddD9W0cAAACgQyazAAAAAOiGMgsAAACAbiizAAAAAOiGMgsAAACAbrgAPAAAADTyknef1DrCIM5/VOsEzDJlFgAAANCXy45pnWAYW1sH6INthgAAAAB0Q5kFAAAAQDdsMwSaOPYBm1tHGM7XWwcAAACYXcosAACmztevP7R1BABgStlmCAAAAEA3lFkAAAAAdEOZBQAAAEA3XDMLlqkcdL/WEQCYZpcd0zrBMLa2DgAArFQmswAAAADohsksAFYU75C2dF9/2cbWEYaxtXUAAAAWw2QWAAAAAN1QZgEAAADQDWUWAAAAAN1wzSwAYCzPfEXrBMPY1joAAACLYjILAAAAgG4oswAAAADohjILAAAAgG4oswAAAADohjILAAAAgG4oswAAAADohjILAAAAgG6sbh0AAACAvm1Zs7F1BGAFMZkFAAAAQDdMZgEADGjLsw9tHQEAYKaYzAIAAACgG8osAAAAALqhzAIAAACgG8osAAAAALqhzAIAAACgG97NEAAA4DsuO6Z1guFsbR0AYDJMZgEAAADQDZNZAAAA0MqsTgNubR2AWWYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6IYyCwAAAIBuKLMAAAAA6Mbq1gEAAAAAFmPLsw9tHYGGTGYBAAAA0A1lFgAAAADdUGYBAAAA0A1lFgAAAADdUGYBAAAA0A3vZggAAMCyXHpq6wTD2dY6AHA7JrMAAAAA6IYyCwAAAIBu2GYIAAAAjWx59qGtI0B3lFl81+t++KGtIwxiS+sAAAAAwMQoswAAAEZMyQBMP9fMAgAAAKAbJrNgmS668W9aRxjElns+rXUEAAAAuB1lFgAAANCVS09tnWAY21oH6MTMlVnHPmBz6wjD+XrrAAAAAABtuWYWAAAAAN1QZgEAAADQDWUWAAAAAN2YuWtmwf521EH3bx0BWIRnvqJ1guG4YCgAACvBzJVZm47e0DrCYHZkR+sIAAD7heKZVmb1HdISzz1gdthmCAAAAEA3lFkAAAAAdEOZBQAAAEA3JnbNrFLKyUnWJblTrfX8PW9P6nEA6JN1AoD5WCcAGFeptU7mQKW8tNZ6ZinlvCTnJDl/99u11l173P+MJGeMbh6d5F8nEgRgttxcaz2xdYhJsE4ADMI6Mcc6AbB3M7NO7G6S72a4c/TxgMxtX9zz9vcsPrXWC5NcOMHHB2C6WScAmI91AoCxTHIy678luV+SeyT5YJI7j24fUGt96UQeBIBuWScAmI91AoBxTazMAgAAAICheTfDMZVSXjr6+IOllGP2+N7WUsrd2yTrw3fOH+PznFs6zzda8JpdHq/bxfOcWzrPN1rwml0er9vF85xbOs+36Wcya0yllNcm+WSSX8rcBSkfmORbSR6Z5BNJbktyVK31V5uFnGKllG1Jvpy5c3Z55s7bHZP8W5IH1Vp/t2G8qeQ5t3Seb7TgNbs8XreL5zm3dJ5vtOA1uzxet4vnObd0nm/Tz2TW+G6ptb48ybdHtw+utV6Q5KbR7Vclua6Ucrcm6abfY5NcluS8JCcmOShzF/V8SOZ+QXB7nnNL5/lGC16zy+N1u3iec0vn+UYLXrPL43W7eJ5zS+f5NuWUWePbtcftb5RSfj3Jvfb4ftl/kbqyPcnPJvntJO/I3C/Qf8/c+fv7drGmmufc0m2P5xv7n9fs8myP1+1iec4t3fZ4vrH/ec0uz/Z43S6W59zSbY/n21SzzXCJSilPTbI6ybdqrdta52H2ec5BX7xm2d8856AvXrPsb55zzBJlFgAAAADdsM1wCUbv/HBgKeX5pZR/GH18YOtc0855m5xSypNKKT88+vy3Sil/V0o5t5TyyNbZppnzxv7i993SOG+T4/fd0jhv7C9+3y2N8zY5ft8tjfM2PVa3DtCxWmt9USnlnrXWF7UO0xHnbR9KKa9M8tEkP5LkK5m7sOA7kjw3yTuT3LnW+so9f67W+vJSyj2SvLTW+o39l3g6OG9MMb/vlsZ52we/75bGeWOK+X23NM7bPvh9tzTOW59MZsH0uK3W+idJfiz/+c4Zj0/ymVrrhUnWllJ+rpRyVsuQU8h5A1YKv++WxnkDVgq/75bGeeuQyayle04p5T+SHNI6SGect337zgXstmfunTNOyty/CPzid+9Q6yXJ3HhrkjNKKV9M8uf7NeX0cd6YVn7fLY3ztm9+3y2N88a08vtuaZy3ffP7bmmctw65ADwAAAAA3bDNEAAAAIBu2Ga4BKWUA5P8tyRrkhye5Kwkp9RaX98y17Tb7bwdk+S9mRvjdN4WsNt5u1uSg5O8Is7bPu3lefaZJE9I8qjMnbujkryu1vofrTIy20bPwd9KckCS65O8OV6zY9nt9XvvJCcnOTHO3YKsE4tjnaA168TS7PbaPSzJXZL8YZy3sfj76+LsZZ14d5JLk1yV5I2ZWzOsE42ZzFqa05JsH717xh2S3JLkvqWUA9rGmnqnZa7A+kiSjN7xwXlb2GlJ/jrJDUnu5rwtaM/n2TVJXpPkmlrrFUnel+SJzdKxEpyW5MgkNyfZ5TW7KN/5fffpJM937sZmnVgc6wStWSeW5rQkx2euBPyc87Yo/v66ON+zTiR5UJIbk9w9yTdinZgKyqylWZfk86WUc5O8tta6K8mXk6xtG2vqrau1Xpfkc7t9zXlb2Lokh9Va35HkDqWUVXHe5rO359kvJXnD6PPPJPnh/Z6KlWRdkq2jt3B+yOh/FL1mx7Ou1npTkhNqrdtHX3PuFmadWBzrBK1ZJ5ZmXZKDa61/kOShftctir+/Ls6e68Snk5yZuYvC/3SsE1PBNsOluSHJHyX5VpKfKKV8LMm9kvzvpqmm3w2llCP2+JrztrAbkvxIKeUnk3yt1rqrlOK87dvenmfraq1/NPr8AUn+dT9nYmW5IclppZRbknyh1vptr9mx3VBKuV/mphW+w7lbmHVicawTtGadWJobkvz/pZQzk3zJ77pF8ffXxdlznTg8yRmZm8q6MNaJqeDdDJeglHLXJL9Ya/2T0e2DkvxUrfUN8//kyua8LY3ztjh7nq+9fP/pSf641rpz/yZjpfCaXTrnbmmct8WxTtCa1+zSOG9L59wtjnWiD8osAAAAALrhmlkAAAAAdEOZBQAAAEA3lFkAAAAAdEOZBQAAAEA3lFkAAAAAdOP/AqoFQ1O2RRyOAAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 1440x1080 with 3 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "fig = plt.figure(figsize=(20,15))\n",
    "\n",
    "ax1 = plt.subplot2grid((1, 3), (0, 0))\n",
    "ax2 = plt.subplot2grid((1, 3), (0, 1))\n",
    "ax3 = plt.subplot2grid((1, 3), (0, 2))\n",
    "\n",
    "width = 0.7\n",
    "\n",
    "\n",
    "####### LUNG(0), STOMACH(1), THROAT(2) ###########\n",
    "\n",
    "axes = [ax1, ax2, ax3]\n",
    "\n",
    "taxa_list_reduced = []\n",
    "\n",
    "for i, df in enumerate(BAplot_df):\n",
    "\n",
    "    order_tmp = list(set(df['type'].tolist()))\n",
    "    order = order_list(order_tmp)\n",
    "\n",
    "    # To get the same color for the taxa each time when plotting \n",
    "    tax_list = []\n",
    "    for x in range(len(list(df.set_index('type').loc[order].groupby(['type', 'Full Taxa'], sort=False).sum().unstack()))):\n",
    "        tax_list.append(list(df.set_index('type').loc[order].groupby(['type', 'Full Taxa'], sort=False).sum().unstack())[x][1])\n",
    "\n",
    "    hh = df.set_index('type').loc[order].groupby(['type', 'Full Taxa'], sort=False).sum().unstack().plot.bar(\n",
    "        stacked=True, color=[test.get(x, '#333333') for x in tax_list], ax=axes[i], width = width, edgecolor = \"k\", linewidth=0)\n",
    "\n",
    "    # To store which taxa are present in the plot\n",
    "    for tax in tax_list:\n",
    "        taxa_list_reduced.append(tax)\n",
    "\n",
    "    axes[i].set_xlabel('')\n",
    "    axes[i].legend().remove()\n",
    "    axes[i].tick_params(axis='x', which= 'both', length=0)\n",
    "    axes[i].tick_params(axis='y', labelsize=6)\n",
    "\n",
    "    # Y axis label\n",
    "    if i == 0:\n",
    "        axes[i].set_ylabel('Relative abundance (%)', fontsize=7)\n",
    "\n",
    "    # X ticks\n",
    "    labels = [item.get_text() for item in axes[i].get_xticklabels()] \n",
    "    newlabels = []\n",
    "    for lab in labels:\n",
    "        if lab.split(' ')[-2] == 'Transplant':\n",
    "            shorter = 'LT'\n",
    "        elif lab.split(' ')[-2] == 'Other':\n",
    "            shorter = 'non-LT'\n",
    "        newname = lab.split(' ')[0] + '\\n' + shorter + '\\n' + lab.split(' ')[-1]\n",
    "        newlabels.append(newname)\n",
    "\n",
    "    axes[i].set_xticklabels(newlabels, fontsize=6, rotation=0)\n",
    "\n",
    "    if i in [2]:\n",
    "        axes[i].text(4, 0.5, 'Total Bile \\nAcid in GF', fontsize=7, ha=\"center\", va=\"top\", \n",
    "                         color='black', bbox=dict(facecolor='none', edgecolor='black'))\n",
    "        \n",
    "    axes[i].spines['top'].set_visible(False)\n",
    "    axes[i].spines['right'].set_visible(False)\n",
    "    axes[i].spines['bottom'].set_color('k')\n",
    "    axes[i].spines['left'].set_color('k') \n",
    "\n",
    "        \n",
    "# Adjusting x and y axis labels on subplot 1 and 2\n",
    "ax1.set_title('BAL', fontsize = 7)\n",
    "ax2.set_title('Gastric fluid', fontsize = 7)\n",
    "ax3.set_title('Oropharynx', fontsize = 7)\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
